pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,345 @@
+	"fmt"
+	"strings"
+
+	"knative.dev/pkg/logging"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	podutil "github.com/aws/karpenter/pkg/utils/pod"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	resourceType     = "resource_type"
+	nodeName         = "node_name"
+	nodeProvisioner  = "provisioner"
+	nodeZone         = "zone"
+	nodeArchitecture = "arch"
+	nodeCapacityType = "capacity_type"
+	nodeInstanceType = "instance_type"
+	nodePhase        = "phase"
+)
+
+var (
+	allocatableGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "allocatable",
+			Help:      "Node allocatable",
+		},
+		labelNames(),
+	)
+	podRequestsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_pod_requests",
+			Help:      "Node total pod requests",
+		},
+		labelNames(),
+	)
+	podLimitsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_pod_limits",
+			Help:      "Node total pod limits",
+		},
+		labelNames(),
+	)
+	daemonRequestsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_daemon_requests",
+			Help:      "Node total daemon requests",
+		},
+		labelNames(),
+	)
+	daemonLimitsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_daemon_limits",
+			Help:      "Node total daemon limits",
+		},
+		labelNames(),
+	)
+	overheadGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "system_overhead",
+			Help:      "Node system daemon overhead",
+		},
+		labelNames(),
+	)
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(allocatableGaugeVec)
+	crmetrics.Registry.MustRegister(podRequestsGaugeVec)
+	crmetrics.Registry.MustRegister(podLimitsGaugeVec)
+	crmetrics.Registry.MustRegister(daemonRequestsGaugeVec)
+	crmetrics.Registry.MustRegister(daemonLimitsGaugeVec)
+	crmetrics.Registry.MustRegister(overheadGaugeVec)
+}
+
+func labelNames() []string {
+	return []string{
+		resourceType,
+		nodeName,
+		nodeProvisioner,
+		nodeZone,
+		nodeArchitecture,
+		nodeCapacityType,
+		nodeInstanceType,
+		nodePhase,
+	}
+}
+
+type Controller struct {
+	KubeClient    client.Client
+	LabelSliceMap map[types.NamespacedName][]prometheus.Labels
```

Comment:
IIUC, go considers naming variables with their type to be a bit of an anti-style. I'd just call this labels.

Reply:
Good point. Done